add out variant of linear

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61824
* #61821
* #61802
* __->__ #61801
* #61800

resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812510](https://our.internmc.facebook.com/intern/diff/D29812510)

makslevental committed:
add out variant of linear and fix adaptive_avg cuda bug

[ghstack-poisoned]

desertfire commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:
```diff
@@ -727,9 +727,9 @@ namespace {
     const Tensor& input,
-    IntArrayRef output_size)
+    IntArrayRef output_size,
+    Tensor& output)
```
Forgot the include your previous fix in native_functions.yaml? Can we make this as a separate PR since you have mastered ghstack? It will make the review and triage easier if anything goes wrong.

makslevental replied:
@desertfire https://github.com/pytorch/pytorch/pull/61851

makslevental committed:
Update on "add out variant of linear and fix adaptive_avg cuda bug"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

[ghstack-poisoned]

Commit:
Update on "add out variant of linear"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

[ghstack-poisoned]

Commit:
Update on "add out variant of linear"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812510).

desertfire commented on test/cpp/api/operations.cpp:
```diff
@@ -55,3 +55,33 @@ TEST_F(OperationTest, Cross) {
   }
 }
+
+TEST_F(OperationTest, Linear_out) {
```
I remember Nick posted some comments about this test, mainly on the integer type. Can we use an approach similar to `TEST_F(OperationTest, Cross)` where we use `torch::rand` to generate the input and use loops to compute the expected result?

makslevental replied:
@desertfire nick added a comment about using `OpInfo` instead of this approach. I agree it's a better approach but I would like to land this one and then submit the `OpInfo` in a separate PR.

makslevental commented:
@desertfire , @hlu1 , @eellison and @Krovatkin  any blockers here?

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812510).

makslevental committed:
Update on "add out variant of linear"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812510](https://our.internmc.facebook.com/intern/diff/D29812510)

[ghstack-poisoned]
